### PR TITLE
common/compiler: fix #1598, expose solidity errors

### LINKED ISSUE #1598: expose Solidity compilation error message
when compiling solidity code, `web3.eth.compile.solidity(source)` doesn't expose the error message; instead it only returns:
```
{"id":4,"jsonrpc":"2.0","error":{"code":-32603,"message":"solc error: exit status 1"}}
```

For development purposes it would be most useful to also return the actual error message from `stderr`; as they are returned when calling the compiler from the CLI:
```
$ solc app/contracts/simple_storage.sol
app/contracts/simple_storage.sol:12:17: Parser error: Expected primary expression.
    storedData =* 2;
                ^
```

### PR DESCRIPTION
**Pull request by karalabe**
This PR cleans up a bit the solidity compilation code + exposes the stderr stream of the solc compiler in case of a build error.

### CODE REVIEW

**Comment by robotally** [comment]
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Tue Aug 25 16:43:31 UTC 2015_

**Comment by karalabe** [comment]
@zelig PTAL

File: common/compiler/solidity.go [hunk 1]
```diff
@@ -110,95 +111,82 @@ func (sol *Solidity) Version() string {
...
+	// Sanity check that something was actually built
 	matches, _ := filepath.Glob(wd + "/*.binary")
 	if len(matches) < 1 {
-		err = fmt.Errorf("solc error: missing code output")
-		return
+		return nil, fmt.Errorf("solc: no build results found")
 	}
-
-	contracts = make(map[string]*Contract)
+	// Complilation succeeded, assemble and return the contracts
```
Comment (fjl) [inline common/compiler/solidity.go]: `// Compilation succeeded, assemble and return the contracts.`

Sentences. End. With. Period.

Comment (karalabe) [inline common/compiler/solidity.go]: I always use full sentences as variable, fumction, etc declarations. Within
the code, I never add the dots. It's a personal style :D

Comment (bas-vk) [inline common/compiler/solidity.go]: typo Complilation -> Compilation

Comment (karalabe) [inline common/compiler/solidity.go]: Btw, if you look through the Go stdlib, they follow the same pattern (e.g. https://golang.org/src/math/sin.go?s=3697:3724#L191). All longer descriptions in front og public entities are full sentences, but internal "notes" are without a dot, so I'd like to keep to this.

Also it's a bit cleaner to me without a dot. The dot doesn't add any value to a few word comment, but both makes it longer as well as (for me) draws attention as a syntax element :P

File: common/compiler/solidity.go [hunk 2]
```diff
@@ -110,95 +111,82 @@ func (sol *Solidity) Version() string {
 	return sol.version
 }
 
-func (sol *Solidity) Compile(source string) (contracts map[string]*Contract, err error) {
-
+// Compile builds and returns all the contracts contained within a source string.
+func (sol *Solidity) Compile(source string) (map[string]*Contract, error) {
+	// Short circuit if no source code was sepcified
```
Comment (bas-vk) [inline common/compiler/solidity.go]: Typo, sepcified -> specified

**Comment by bas-vk** [comment]
The solc error doesn't seem to be parsed and displayed in the console.

```
> eth.compile.solidity("contract a{)")
Invalid JSON RPC response: undefined
    at InvalidResponse (<anonymous>:-81076:-37)
    at send (<anonymous>:-154580:-37)
    at solidity (<anonymous>:-131712:-37)
    at <anonymous>:1:1
```
It don't think it's related to this PR and *might* have something to do with https://github.com/ethereum/go-ethereum/issues/1709. I'm investigating this.

**Comment by karalabe** [comment]
@bas-vk @fjl Fixed the two typos. Still struggling wit the atom spell checker... it doesn't support looking inside comments only so it's very annoying... if you know a fix for this? :D

**Comment by karalabe** [comment]
@bas-vk Maybe these are separate issues. Running the embark framework on your invalid "file" results in:

```
Warning: solc: exit status 1
<stdin>:1:12: Parser error: Function, variable, struct or modifier declaration expected.
contract a{)
           ^
 Use --force to continue.
```

**Comment by bas-vk** [comment]
There is a bug in the console error handling/parsing. In case an error contains specific characters the `JSON.parse` method fails. I'm not familiar with embark but I guess they are using the http interface and don't use the `Jeth.err` method. I will create a PR for that.

For this PR :+1:

**Comment by fjl** [comment]
:+1:

**Comment by zelig** [comment]
:+1: perfetto maestro

**Pull request merged**